api/singlesignon.go

```diff
@@ -0,0 +1,116 @@
+package api
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/crewjam/saml"
+	"github.com/gofrs/uuid"
+	"github.com/netlify/gotrue/models"
+	"github.com/netlify/gotrue/storage"
+	"github.com/netlify/gotrue/utilities"
+)
+
+type SingleSignOnParams struct {
+	ProviderID uuid.UUID `json:"provider_id"`
+	Domain     string    `json:"domain"`
+	RedirectTo string    `json:"redirect_to"`
+}
+
+func (p *SingleSignOnParams) validate() error {
+	if p.ProviderID == uuid.Nil && p.Domain == "" {
+		return badRequestError("Only provider_id or domain supported")
+	} else if p.ProviderID != uuid.Nil && p.Domain != "" {
+		return badRequestError("A provider_id or domain needs to be provided")
+	}
+
+	return nil
+}
+
+// SingleSignOn handles the single-sign-on flow for a provided SSO domain or provider.
+func (a *API) SingleSignOn(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	db := a.db.WithContext(ctx)
+
+	body, err := getBodyBytes(r)
+	if err != nil {
+		return internalServerError("Unable to read request body").WithInternalError(err)
+	}
+
+	var params SingleSignOnParams
+
+	if err := json.Unmarshal(body, &params); err != nil {
+		return badRequestError("Unable to parse request body as JSON").WithInternalError(err)
+	}
+
+	if err := params.validate(); err != nil {
+		return err
+	}
+
+	var ssoProvider *models.SSOProvider
+
+	if params.ProviderID != uuid.Nil {
+		ssoProvider, err = models.FindSSOProviderByID(db, params.ProviderID)
+		if models.IsNotFoundError(err) {
+			return notFoundError("No such SSO provider")
+		} else if err != nil {
+			return internalServerError("Unable to find SSO provider by ID").WithInternalError(err)
+		}
+	} else if params.Domain != "" {
+		ssoProvider, err = models.FindSSOProviderByDomain(db, params.Domain)
+		if models.IsNotFoundError(err) {
+			return notFoundError("No SSO provider assigned for this domain")
+		} else if err != nil {
+			return internalServerError("Unable to find SSO provider by domain").WithInternalError(err)
+		}
+	}
+
+	if ssoProvider == nil {
+		// impossible condition since #validate() should have handled this case
+		return internalServerError("Unable to find SSO provider")
+	}
```

Comment:
nit: might want to consider making `validate()` return `(hasProviderId bool, err error)` so you don't have to check `params.ProviderID != uuid.Nil` and `params.Domain != ""` again

Reply:
and you can do away with the LOC with the impossible condition clause